crates/typst-macros/src/elem.rs

```diff
@@ -480,88 +472,904 @@ fn create_param_info(field: &Field) -> TokenStream {
+            #[doc = #doc]
+            #vis fn #ident_in<'a: 'b, 'b>(styles: &'b ::typst::model::StyleChain<'a>) -> ::std::borrow::Cow<'b, #output> {
+                #access
+            }
+        }
+    } else {
+        quote! {
+            #[doc = #doc]
+            #vis fn #ident_in(styles: ::typst::model::StyleChain) -> #output {
+                #access
+            }
+        }
+    }
+}
+
+/// Create an accessor methods for a field.
+fn create_field_method(element: &Elem, field: &Field) -> TokenStream {
+    let Field { vis, docs, ident, output, .. } = field;
+    if field.inherent() && !field.synthesized {
+        quote! {
+            #[doc = #docs]
+            #[track_caller]
+            #vis fn #ident(&self) -> &#output {
+                &self.#ident
+            }
+        }
+    } else if field.synthesized {
+        quote! {
+            #[doc = #docs]
+            #[track_caller]
+            #vis fn #ident(&self) -> &#output {
+                self.#ident.as_ref().unwrap()
+            }
+        }
+    } else if field.borrowed {
+        let access =
+            create_style_chain_access(element, field, quote! { self.#ident.as_ref() });
+
+        quote! {
+            #[doc = #docs]
+            #vis fn #ident<'a: 'b, 'b>(&'b self, styles: &'b ::typst::model::StyleChain<'a>) -> ::std::borrow::Cow<'b, #output> {
+                #access
+            }
+        }
+    } else {
+        let access =
+            create_style_chain_access(element, field, quote! { self.#ident.as_ref() });
+
+        quote! {
+            #[doc = #docs]
+            #vis fn #ident(&self, styles: ::typst::model::StyleChain) -> #output {
+                #access
+            }
+        }
+    }
+}
+
+/// Create a style chain access method for a field.
+fn create_style_chain_access(
+    element: &Elem,
+    field: &Field,
+    inherent: TokenStream,
+) -> TokenStream {
+    let enum_ = &element.enum_ident;
+    let Field { ty, default, enum_ident, .. } = field;
+    let getter = match (field.fold, field.resolve, field.borrowed) {
+        (false, false, false) => quote! { get },
+        (false, false, true) => quote! { get_borrowed },
+        (false, true, _) => quote! { get_resolve },
+        (true, false, _) => quote! { get_fold },
+        (true, true, _) => quote! { get_resolve_fold },
+    };
+
+    quote! {
+        styles.#getter::<#ty>(
+            <Self as ::typst::model::NativeElement>::elem(),
+            #enum_::#enum_ident,
+            #inherent,
+            || #default,
+        )
+    }
+}
+
+/// Create a field declaration.
+fn create_field(field: &Field) -> TokenStream {
+    let Field { ident, ty, docs, required, .. } = field;
+
+    let ty = required
+        .then(|| quote! { #ty })
+        .unwrap_or_else(|| quote! { Option<#ty> });
+    quote! {
+        #[doc = #docs]
+        #ident: #ty
+    }
+}
+
+/// Creates the element's `Hash` implementation.
+fn create_hash_impl(element: &Elem) -> TokenStream {
```

Comment:
why can't we just derive hash?

Reply:
This implementation mimicks the way the old structure worked by allowing `Hash` to skip fields that are not set. I initially did this for compatibility with the old system (when there were both in) and for the `#[not_hash]` attribute, but now that it's gone I suppose there is no reason not to just derive it.